Use SearchResult component

tsmacdonald opened the pull request:
...and update the backend to return much richer match data

[Fixes #14832]


![image](https://user-images.githubusercontent.com/784417/108543205-633ed100-72aa-11eb-8fc0-7383baa90411.png)

tsmacdonald committed:
Use SearchResult component

Update the backend to return much richer match data

[Fixes #14832]

tsmacdonald commented on src/metabase/search/scoring.clj:
```diff
@@ -117,6 +133,18 @@
+       :name           (if (or (= column :name) (nil? display_name)) name display_name)
+       :matched_column column
+       :matched_text   match
+       :context        (when-not (#{:name :display_name :collection_name} column) match-context) ;; TODO pull this out somewhere more responsible
```
these "magic" column names should live in `search.config` and both the `when-not` and `if` (L141) are a little jank, but I think the surrounding logic is going to change in the near future and don't want to waste cycles overoptimizing

kdoh commented on frontend/src/metabase/search/components/SearchResult.jsx:
```diff
@@ -62,6 +62,8 @@ export default function SearchResult({ result }) {
     case "collection":
       return <CollectionResult collection={result} />;
+    case "table":
+      return <TableResult table={result} />;
```
It looks like we're not accounting for segments or metrics here yet.

kdoh reviewed (changes requested):
Looks like a few things are missing before this can go in and we can test further.

1. We should use SearchResult on both the full search results page and in type ahead results.
2. The `Our analytics` collection isn't showing up as the collection when something is saved there. I think that's due to the back end treating that collection strangely.
3. I think Metrics and Segments might not be accounted for yet which also tells me we need to improve our fallback case.

tsmacdonald committed:
Format metrics and segments in search results; DRY up the code

tsmacdonald committed:
Increase the memoization cache for search scoring by common subseq (#14894)

* Increase the memoization cache for search scoring by common subseq

Also limit the size of input it can get

(Hopefully) [Fixes #14852]

* Don't group search results by type (#14898)

Respect the sorting done by the backend

tsmacdonald committed:
Add normalizedCollection(); show "Our Analytics" collection in search results

tsmacdonald committed:
Use SearchResult for typeahead results too

tsmacdonald commented on frontend/src/metabase/entities/collections.js:
```diff
@@ -151,6 +151,14 @@ export const canonicalCollectionId = (
+  }
+  return collection;
+}
+
```
I believe it both accepts and returns an `ExpandedCollection`, but I couldn't get Flow to be happy and since we're planning to remove it pretty soon anyway I gave up.

paulrosenzweig replied:
Maybe this could call `canonicalCollectionId` on the id here? I'm not sure how a collection with `id: "root"` that isn't `ROOT_COLLECTION` would end up here but it might be good to handle since I'm wary of the somewhat messy canonicalization.

Punting on trying to make Flow happy makes sense to me.

tsmacdonald replied:
Fair point; fixed!
